utils and libraries: Replace deprecated Boost Filesystem functions

hebasto opened the pull request:
Boost Filesystem `basename()` and `extension()` functions are [deprecated since v1.36.0](https://www.boost.org/doc/libs/1_36_0/libs/filesystem/doc/reference.html#Convenience-functions).

See more: https://lists.boost.org/Archives/boost/2010/01/160905.php

Also this PR prevents further use of deprecated Boost Filesystem functions.
Ref: https://www.boost.org/doc/libs/1_64_0/libs/filesystem/doc/index.htm#Coding-guidelines

Note: On my Linux system Boost 1.65.1 header `/usr/include/boost/filesystem/convenience.hpp` contains:
```c++
# ifndef BOOST_FILESYSTEM_NO_DEPRECATED

    inline std::string extension(const path & p)
    {
      return p.extension().string();
    }

    inline std::string basename(const path & p)
    {
      return p.stem().string();
    }

    inline path change_extension( const path & p, const path & new_extension )
    { 
      path new_p( p );
      new_p.replace_extension( new_extension );
      return new_p;
    }

# endif
```

UPDATE:
Also removed unused code as [noted](https://github.com/bitcoin/bitcoin/pull/15880#discussion_r279386614) by **ryanofsky**.

Empact commented on src/wallet/db.cpp:
```diff
@@ -408,8 +408,8 @@ bool BerkeleyBatch::VerifyEnvironment(const fs::path& file_path, std::string& er
     // Wallet file must be a plain filename without a directory
-    if (walletFile != fs::basename(walletFile) + fs::extension(walletFile))
-    {
+    fs::path wallet_file_path = fs::path(walletFile);
```
nit: could construct directly

hebasto replied:
@Empact Thank you for your review.
I'm not sure if I understand your comment correctly. Could you elaborate, please?

Empact replied:
```diff
-fs::path wallet_file_path = fs::path(walletFile);
+fs::path wallet_file_path(walletFile);
```

hebasto replied:
Thank you! Fixed.

ryanofsky commented on src/wallet/db.cpp:
```diff
@@ -408,8 +408,8 @@ bool BerkeleyBatch::VerifyEnvironment(const fs::path& file_path, std::string& er
-    if (walletFile != fs::basename(walletFile) + fs::extension(walletFile))
-    {
+    fs::path wallet_file_path(walletFile);
+    if (walletFile != wallet_file_path.stem().string() + wallet_file_path.extension().string()) {
```
Should just drop this check. The way `SplitWalletPath` is defined this can never be false:

https://github.com/bitcoin/bitcoin/blob/8da1aa471eb9f062c99c3134d6151961619e9da5/src/wallet/db.cpp#L66

Also, the way `stem()` and `extension()` are defined you could replace them with a call to `filename()`

```c++
if (walletFile != wallet_file_path.filename().string()) {
```

https://www.boost.org/doc/libs/1_69_0/libs/filesystem/doc/reference.html#path-decomposition

ryanofsky reviewed (approved):
utACK 8823a11988d0d4824a44eb2f34e37632e7a0f09a

hebasto committed:
Remove dead code for walletFile check

SplitWalletPath() garanties the walletFile is a plain filename without a
directory.

hebasto committed:
Replace deprecated Boost Filesystem function

Boost Filesystem basename() function is deprecated since v1.36.0.
Also, defining BOOST_FILESYSTEM_NO_DEPRECATED before including
filesystem headers is strongly recommended. This prevents inadvertent
use of old features, particularly legacy function names, that have been
replaced and are going to go away in the future.

hebasto commented:
@ryanofsky 
Thank you for your review. Your comment has been addressed.
Would you mind re-reviewing?

ryanofsky reviewed (approved):
utACK 276a67490f795429c1af844966ad7960908576c0. Only change is replacing impossible to reach condition with an assert.

promag commented on src/wallet/db.cpp:
```diff
@@ -408,11 +408,7 @@ bool BerkeleyBatch::VerifyEnvironment(const fs::path& file_path, std::string& er
-        errorStr = strprintf(_("Wallet %s resides outside wallet directory %s"), walletFile, walletDir.string());
-        return false;
-    }
+    assert(walletFile == fs::path(walletFile).filename().string());
```
Suggestion, either drop the assertion (`GetWalletEnv` always returns a plain filename) or make it complete because:
```cpp
std::cout << path("/").filename();            // outputs "/"
std::cout << path(".").filename();            // outputs "."
std::cout << path("..").filename();           // outputs ".."
```
so for these cases the assertion passes.

promag reviewed (commented):
utACK 276a674.

Shouldn't 3edc71674ce79d81f4c5da94663787899f123719 be after 276a67490f795429c1af844966ad7960908576c0?

hebasto commented:
@promag thank you for your review.

Your comments have been addressed:
- the assertion has been dropped
- commits have been rearranged

promag commented:
utACK 8d0add4.

Empact commented:
utACK https://github.com/bitcoin/bitcoin/pull/15880/commits/8d0add4ea92b148e5ab8295e9fa9b6776db38c18 could squash the latter two

The test was introduced in #7691 fyi, seems fine to remove.

hebasto commented:
@Empact 
> utACK [8d0add4](https://github.com/bitcoin/bitcoin/commit/8d0add4ea92b148e5ab8295e9fa9b6776db38c18) could squash the latter two

Done.

> The test was introduced in #7691 fyi, seems fine to remove.

I guess it was earlier in #1889 ;)

Empact commented:
utACK https://github.com/bitcoin/bitcoin/pull/15880/commits/a0a222eec0b7f615a756e5e0dcec9b02296f999c

> I guess it was earlier in #1889 ;)

Indeed! :P

ryanofsky reviewed (approved):
utACK a0a222eec0b7f615a756e5e0dcec9b02296f999c. Only change is dropping assert and squashing first two commits.

practicalswift commented:
utACK a0a222eec0b7f615a756e5e0dcec9b02296f999c

fanquake commented:
utACK a0a222e
